### PR TITLE
Create `rustc_type_ir`

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Decided to start small 😄

This PR creates a `rustc_type_ir` crate as part of the WG-Traits plan to create a shared type library.
~~There already exists a `rustc_ty` crate, so I named the new crate `rustc_ty_library`. However I think it would make sense to rename the current `rustc_ty` to something else (e.g. `rustc_ty_passes`) to free the name for this new crate.~~

r? @jackh726

### CODE REVIEW

**Comment by jackh726** [comment]
I haven't read through this yet, but I agree that we should rename `rustc_ty` to `rustc_ty_passes` (or similiar), and probably prior to this. Maybe make a separate PR. *This* is probably MCP territory, but I think the rename would probably be fine. Though we'll need @rust-lang/compiler signoff.

**Review by jackh726 — APPROVED** [review]
This LGTM! I'm on board with these changes as-is (only a fairly small subset of what will eventually get moved out), but I guess there are couple questions/concerns others might have:
1) Should we be moving out *more*? I think *no* since we've discussed this in wg-traits meetings and decided any start is betting than nothing.
2) Should things still have a shorthand, like `ty::INNERMOST` to `tylib::INNERMOST` or something like that?
3) Should this wait until after `rust_ty` MCP? Should this have an MCP itself? I think it's best to probably discuss this on Zulip.

And finally, let's r? @nikomatsakis, since I don't have r+ rights.

**Comment by jackh726** [comment]
>Should things still have a shorthand, like ty::INNERMOST to tylib::INNERMOST or something like that?

Another alternative would be to just make `rustc_middle` reexport `rustc_ty_library` items.

**Comment by jyn514** [comment]
`library` is not really descriptive of what this does. Could you call it `rustc_ty_data` instead?

File: compiler/rustc_data_structures/Cargo.toml
Comment (cjgillot) [inline compiler/rustc_data_structures/Cargo.toml:27]: This dependency will cause problems when moving more elaborate data structures to `rustc_ty_library`. Having `rustc_ty_library` depend on `rustc_data_structures` may be more robust.

Comment (jackh726) [reply compiler/rustc_data_structures/Cargo.toml:27]: Eek. `rustc_data_structures` has a lot of depdencies, most of which we probably don't need/want for the eventual type library. Wondering if it's worth thinking about splitting `rustc_data_structures` or just giving in to the extra dependencies.

Granted, this sort of thing was something we expected to encounter rather quickly when starting this process, so it's not surprising at all.

**Review by jackh726 — APPROVED** [review]
This is a good start. Simple, but good.

**Review by nikomatsakis — APPROVED** [review]
r=me

File: compiler/rustc_middle/src/ty/sty.rs
Comment (lqd) [inline compiler/rustc_middle/src/ty/sty.rs:1319]: just a note, and this is pre-existing of course, but reading the diff made me notice `arrays` should be `arrows` here

Comment (LeSeulArtichaut) [reply compiler/rustc_middle/src/ty/sty.rs:1319]: I'll file a PR once this is merged

**Comment by rylev** [comment]
@LeSeulArtichaut @nikomatsakis this PR had some [performance regressions](https://perf.rust-lang.org/compare.html?start=c3ed6681ff8d446e68ce272be4bf66f4145f6e29&end=3f2088aa603d2cd3f43c20795872de9cd6ec7735&stat=instructions:u) in the const evaluation stress test (but not in other benchmarks). Is this change simply moving code around? I wonder if something that was previously being inlined is now not.

**Comment by jyn514** [comment]
Self-profile shows almost all the time being spent in eval_to_allocation_raw:

Screenshot_20201215-111057_Ecosia

**Comment by jackh726** [comment]
> @LeSeulArtichaut @nikomatsakis this PR had some [performance regressions](https://perf.rust-lang.org/compare.html?start=c3ed6681ff8d446e68ce272be4bf66f4145f6e29&end=3f2088aa603d2cd3f43c20795872de9cd6ec7735&stat=instructions:u) in the const evaluation stress test (but not in other benchmarks). Is this change simply moving code around? I wonder if something that was previously being inlined is now not.

Yes, this is literally just moving code around. Didn't expect a perf regression at all. I do imagine something isn't getting inlined anymore.

**Comment by LeSeulArtichaut** [comment]
Filed #80057 to add `#[inline]` to the `DebruijnIndex` methods

**Comment by mati865** [comment]
> Yes, this is literally just moving code around. Didn't expect a perf regression at all.

It's fairly common that PRs moving the code to a new crate regress performance a little.

> I do imagine something isn't getting inlined anymore.

That's often the reason.

**Pull request merged**
